engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraint.java

```diff
@@ -46,22 +47,23 @@
 	private final ConstraintLocation location;
 
 	/**
-	 * The {@link ValueExtractor} used to extract the value for validation.
+	 * The sequence of {@link ValueExtractor}s used to navigate from the outermost container to the innermost container
+	 * and extract the value for validation.
 	 */
-	private final ValueExtractorDescriptor valueExtractorDescriptor;
+	private final List<ValueExtractorDescriptor> valueExtractorDescriptors;
```

Comment:
Yeah, well, I tried to do that and in the end I thought the best solution was to use `Collections.singletonList`. I introduced a new helper called `CollectionHelper.toImmutableList` (same for `Set`) that we should use to make a list immutable and that make some naive optimization. See this commit https://github.com/hibernate/hibernate-validator/pull/659/commits/662003de38b0428a7a13be68b0d30bf50ea0ec21 .

I think it should be generalized to all the immutable list/set in the metadata. Tell me what you think about it.